File: Src/java/engine/src/test/java/org/opencds/cqf/cql/engine/execution/ExpandValueSetTest.java
package org.opencds.cqf.cql.engine.execution;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.opencds.cqf.cql.engine.execution.CqlConceptTest.assertEqual;

import java.util.Collections;
import java.util.List;
import org.cqframework.cql.cql2elm.LibraryManager;
import org.cqframework.cql.cql2elm.ModelManager;
import org.junit.jupiter.api.Test;
import org.opencds.cqf.cql.engine.runtime.Code;
import org.opencds.cqf.cql.engine.terminology.CodeSystemInfo;
import org.opencds.cqf.cql.engine.terminology.TerminologyProvider;
import org.opencds.cqf.cql.engine.terminology.ValueSetInfo;

class ExpandValueSetTest {

    @Test
    void all_expand_valueset() {
        LibraryManager libraryManager = new LibraryManager(new ModelManager());
        libraryManager.getLibrarySourceLoader().registerProvider(new TestLibrarySourceProvider());

        Code expected = new Code().withCode("M").withSystem("http://test.com/system");
        TerminologyProvider terminologyProvider = new TerminologyProvider() {
            public boolean in(Code code, ValueSetInfo valueSet) {
                return true;
            }

            public Iterable<Code> expand(ValueSetInfo valueSet) {
                return Collections.singletonList(expected);
            }

            public Code lookup(Code code, CodeSystemInfo codeSystem) {
                return null;
            }
        };

        Environment environment = new Environment(libraryManager, null, terminologyProvider);

        CqlEngine engine = new CqlEngine(environment);
        var results = engine.evaluate(CqlTestBase.toElmIdentifier("ExpandValueSetTest"));

        @SuppressWarnings("unchecked")
        List<Code> actual = (List<Code>) results.forExpression("ExpandValueSet").value();
        assertNotNull(actual);
        assertEquals(1, actual.size());

        assertEqual(expected, actual.get(0));
    }
}

File: Src/java/engine/src/jvmTest/kotlin/org/opencds/cqf/cql/engine/execution/ExpandValueSetTest.kt
package org.opencds.cqf.cql.engine.execution

import org.cqframework.cql.cql2elm.LibraryManager
import org.cqframework.cql.cql2elm.ModelManager
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.opencds.cqf.cql.engine.runtime.Code
import org.opencds.cqf.cql.engine.runtime.CqlType
import org.opencds.cqf.cql.engine.terminology.CodeSystemInfo
import org.opencds.cqf.cql.engine.terminology.TerminologyProvider
import org.opencds.cqf.cql.engine.terminology.ValueSetInfo

internal class ExpandValueSetTest {
    @Test
    fun all_expand_valueset() {
        val libraryManager = LibraryManager(ModelManager())
        libraryManager.librarySourceLoader.registerProvider(TestLibrarySourceProvider())

        val expected = Code().withCode("M").withSystem("http://test.com/system")
        val terminologyProvider: TerminologyProvider =
            object : TerminologyProvider {
                override fun `in`(code: Code, valueSet: ValueSetInfo): Boolean {
                    return true
                }

                override fun expand(valueSet: ValueSetInfo): Iterable<Code> {
                    return mutableListOf(expected)
                }

                override fun lookup(code: Code, codeSystem: CodeSystemInfo): Code {
                    error("Not implemented")
                }
            }

        val environment = Environment(libraryManager, null, terminologyProvider)

        val engine = CqlEngine(environment)
        val results = engine.evaluate { library("ExpandValueSetTest") }.onlyResultOrThrow

        val actual = results["ExpandValueSet"]!!.value as MutableList<*>
        Assertions.assertEquals(1, actual.size)

        CqlConceptTest.assertEqual(expected, actual[0] as CqlType)
    }
}
